From patchwork Wed Aug 12 08:28:09 2026
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Mohamed Mediouni <mohamed@unpredictable.fr>
X-Patchwork-Id: 2284252
Return-Path: <qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org>
X-Original-To: incoming@patchwork.ozlabs.org
Delivered-To: patchwork-incoming@legolas.ozlabs.org
Authentication-Results: legolas.ozlabs.org;
	dkim=pass (2048-bit key;
 unprotected) header.d=unpredictable.fr header.i=@unpredictable.fr
 header.a=rsa-sha256 header.s=sig1 header.b=vQNY/NMO;
	dkim-atps=neutral
Authentication-Results: legolas.ozlabs.org;
 spf=pass (sender SPF authorized) smtp.mailfrom=nongnu.org
 (client-ip=209.51.188.17; helo=lists1p.gnu.org;
 envelope-from=qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org;
 receiver=patchwork.ozlabs.org)
Received: from lists1p.gnu.org (lists1p.gnu.org [209.51.188.17])
	(using TLSv1.2 with cipher ECDHE-ECDSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by legolas.ozlabs.org (Postfix) with ESMTPS id 4hKhW34Vlyz1xwH
	for <incoming@patchwork.ozlabs.org>; Wed, 12 Aug 2026 18:29:51 +1000 (AEST)
Received: from localhost ([::1] helo=lists1p.gnu.org)
	by lists1p.gnu.org with esmtp (Exim 4.90_1)
	(envelope-from <qemu-devel-bounces@nongnu.org>)
	id 1wu4K3-0007Vh-Mw; Wed, 12 Aug 2026 04:28:27 -0400
Received: from eggs.gnu.org ([2001:470:142:3::10])
 by lists1p.gnu.org with esmtps (TLS1.2:ECDHE_RSA_AES_256_GCM_SHA384:256)
 (Exim 4.90_1) (envelope-from <mohamed@unpredictable.fr>)
 id 1wu4K2-0007Uz-8j
 for qemu-devel@nongnu.org; Wed, 12 Aug 2026 04:28:26 -0400
Received: from pv-2003i-snip6-3.eps.apple.com ([2a01:b747:3003:205::5d]
 helo=outbound.pv.icloud.com)
 by eggs.gnu.org with esmtps (TLS1.2:ECDHE_RSA_AES_256_GCM_SHA384:256)
 (Exim 4.90_1) (envelope-from <mohamed@unpredictable.fr>)
 id 1wu4K0-00052E-Id
 for qemu-devel@nongnu.org; Wed, 12 Aug 2026 04:28:25 -0400
Received: from outbound.pv.icloud.com (unknown [127.0.0.2])
 by p00-icloudmta-asmtp-us-west-1a-100-percent-10 (Postfix) with ESMTPS id
 CF7641800965; Wed, 12 Aug 2026 08:28:20 +0000 (UTC)
X-ICL-RepId: 019ff516-3396-7b66-a3c6-930e15ee08f6
X-ICL-Out-Info: 
 HUtFAUMHWwJACUgBTUQeDx5WFlZNRAJCTQ1OHVwPUhxEDlYCVRcOVk1UF1sMWg5WMEwWQx9SD1sTTRlRAVJFVAIZHldQQgVCAlwcWg5DClAcXFZVHxcZUU0PWFsIWwQPH0wMUQJCBVZeSQ4dBFQHXQVdVlACWktCBEtFaFwFXBxAF0gdX2pLVhQEDV0dRQ5WGVoMUg9bDhwWS1heBFNWDkIJSgVdD1oAQAFKC1wDRQBHC0kEQAVGXx9RGV4IUytHHkkKVgleCEYRWxRWQ1EZDFBNAUMICghHA00XXjJTBF8RUBY=
Dkim-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=unpredictable.fr;
 s=sig1; t=1786523302; x=1789115302;
 bh=pEhb8ZE16BlOGPNjQ1XeHM1sRdquNH7mB6HdHyGZ8c0=;
 h=From:To:Subject:Date:Message-ID:MIME-Version:x-icloud-hme;
 b=vQNY/NMOBxZZIUy6NzE7G6eh7R70T48yv0hJxFFr10ClNCy0txU1odR28bTOdq8uljMWgBPM+dwtrDioB9ueKWXfnSxNFp169+jZg/YPIxdls+qVK2W3UFFiO34FpGboBmHPNntwtBbf4PtJQMFJGpPCXthTxzaIOWO2AcRpxct/2VX5Muh4X42QDwuUhz4gGpcfFe/LvN5clggSbybYbuPpbUTrwzDTd5MsysJlPvcZsx5jrtUC87Oy323w/L+wvNrwK0DlXEcuu5FJI97rQ4RR0lwnDFRbZXkTxrDj4zhSPKLji/D6ymcTG97HVYeBV3hoU/ol88kZ76Lfi60u3g==
mail-alias-created-date: 1752046281608
Received: from localhost.localdomain (unknown [17.56.9.36])
 by p00-icloudmta-asmtp-us-west-1a-100-percent-10 (Postfix) with ESMTPSA id
 1EFD81800957; Wed, 12 Aug 2026 08:28:18 +0000 (UTC)
From: Mohamed Mediouni <mohamed@unpredictable.fr>
To: qemu-devel@nongnu.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 "Michael S. Tsirkin" <mst@redhat.com>,
 Pedro Barbuda <pbarbuda@microsoft.com>,
 Mohamed Mediouni <mohamed@unpredictable.fr>,
 Richard Henderson <richard.henderson@linaro.org>
Subject: [PATCH 1/6] whpx: i386: fix xsaves enablement in legacy probing path
Date: Wed, 12 Aug 2026 10:28:09 +0200
Message-ID: <20260812082814.27217-2-mohamed@unpredictable.fr>
X-Mailer: git-send-email 2.54.0
In-Reply-To: <20260812082814.27217-1-mohamed@unpredictable.fr>
References: <20260812082814.27217-1-mohamed@unpredictable.fr>
MIME-Version: 1.0
X-Proofpoint-Spam-Details-Enc: AW1haW4tMjYwODEyMDA2NyBTYWx0ZWRfX2JZtRwZ6mn03
 LZaftKkteLJjsN8hPDF4aB54jZb7XqW6pj5inZEhpoHUbHLeyrhZcKwOprihOQsxzBTI5epX6UJ
 MQxL0BYpDqdXQ3tas+Aoor9HZ1moikGipW6OCP5sWQ1JusYJqymmYG1v/GSyBT9BLvCBnnCMXOu
 +fe/v3YMDQpyDSaEz3lGIkIEaa7y/dtTMRwWWP6b5NYR1cIBSwlJl4vYAq/6ehbHUsKlE6uqsey
 znbuO+jxj9oPFi3Fs/JuUzI27PPsM9et3XhtWvd+UFg7Sft0Cmh0rZFPWLdizlSZT+SRN+n4hlX
 RNkOafKlzsfQdhMxmhra17zwk3OZDAymunmxfR5HN0mO8yI4w6j9QmT29NKEXg=
X-Proofpoint-GUID: _dCJ3M6-07YxGyY28wPStP8AwI-y7v3j
X-Proofpoint-ORIG-GUID: _dCJ3M6-07YxGyY28wPStP8AwI-y7v3j
X-Authority-Info-Out: v=2.4 cv=LuOfC3dc c=1 sm=1 tr=0 ts=6a7c2ea5
 cx=c_apl:c_pps:t_out a=azHRBMxVc17uSn+fyuI/eg==:117
 a=azHRBMxVc17uSn+fyuI/eg==:17 a=Sv0fKeRqtYgA:10 a=VkNPw1HP01LnGYTKEx00:22
 a=1u9Qq8vgubJVBI3kFjUA:9
X-JNJ: 
 AAAAAAABIwDISLh1QCXYTGAdhZXLxpzsQBB7vF7rtIJf/r71i/PqhR4XXNiDv7CDPkfgzi9IaMKCnKS0i38yjZ3LTVGkIvTN1kxeQWxW65AAZIUVROjujpjR6pRWuL9/D1Po65I4DpkTbTh9CRWlOLZvLpkrXAlIDV4QGDSkt0Pk3oK3XKAKG0nWoeYmN9tFfm/H8tvk6c4BDeO8qLpKHPeHRcAJH6nJsQLDbarPR+wDjdTrsr/K3bXzd52ZSP6uO0c7WzLQqr70tIKU2aAiT4UybMEpuqD7IjmY7NCmtyELzccCbHgdBASlkVtBteSERuNJOBm32P7cf71NEJZFmhFvqu8zgCTg8i9Dd0Ir38HDfO2L1YWsvjZIrtWe8f0g/Vz3vq/P/TlJ03KiV+ApkQy+gyHZDOf5XqERBG2jZeWdnnGPQycW6/3v8u0OP9v8ZfIWmxMgn77UprciwnBZCdLBQIloPm/B3VvPX7HiubZSPyE4apeHyKVcQ0G4lTHC9TxdpRKpgUt0ja7goDQCZ/1mS0FJ0UehzHX3UlvWqWgUMC7PsU50bst5IuG005sNp5x2Ah/S7XXuVx2RszTxH4FoMHKv5nB373RCXfIhd8E7bEaaSv51IdqETcVeyngv8QO9JKmgWF/YOCMvg3Uuu973cHVCUopC1jSJtR+YCJjz8a324e0u85TpSikHEiPoyfeV+90JlFbf8UHP5a0qdn4sZ7cTPwDtm2Q=
Received-SPF: pass client-ip=2a01:b747:3003:205::5d;
 envelope-from=mohamed@unpredictable.fr; helo=outbound.pv.icloud.com
X-Spam_score_int: -20
X-Spam_score: -2.1
X-Spam_bar: --
X-Spam_report: (-2.1 / 5.0 requ) BAYES_00=-1.9, DKIM_SIGNED=0.1,
 DKIM_VALID=-0.1, DKIM_VALID_AU=-0.1, DKIM_VALID_EF=-0.1,
 SPF_HELO_PASS=-0.001,
 SPF_PASS=-0.001 autolearn=ham autolearn_force=no
X-Spam_action: no action
X-BeenThere: qemu-devel@nongnu.org
X-Mailman-Version: 2.1.29
Precedence: list
List-Id: qemu development <qemu-devel.nongnu.org>
List-Unsubscribe: <https://lists.nongnu.org/mailman/options/qemu-devel>,
 <mailto:qemu-devel-request@nongnu.org?subject=unsubscribe>
List-Archive: <https://lists.nongnu.org/archive/html/qemu-devel>
List-Post: <mailto:qemu-devel@nongnu.org>
List-Help: <mailto:qemu-devel-request@nongnu.org?subject=help>
List-Subscribe: <https://lists.nongnu.org/mailman/listinfo/qemu-devel>,
 <mailto:qemu-devel-request@nongnu.org?subject=subscribe>
Errors-To: qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org
Sender: qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org

Without this, Linux will crash in a configuration where
xsaves is expected.

Signed-off-by: Mohamed Mediouni <mohamed@unpredictable.fr>
---
 target/i386/whpx/whpx-cpu-legacy.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/target/i386/whpx/whpx-cpu-legacy.c b/target/i386/whpx/whpx-cpu-legacy.c
index d341e6f4fd..3fb5ed79e3 100644
--- a/target/i386/whpx/whpx-cpu-legacy.c
+++ b/target/i386/whpx/whpx-cpu-legacy.c
@@ -113,7 +113,7 @@ uint32_t whpx_get_supported_cpuid_legacy(uint32_t func, uint32_t idx,
         if (idx == 0) {
             eax = supported_xcr0;
         } else if (idx == 1) {
-            eax &= CPUID_XSAVE_XSAVEOPT | CPUID_XSAVE_XGETBV1;
+            eax &= CPUID_XSAVE_XSAVEOPT | CPUID_XSAVE_XGETBV1 | CPUID_XSAVE_XSAVES;
             if (!whpx_has_xsaves()) {
                 eax &= ~CPUID_XSAVE_XSAVES;
             }

From patchwork Wed Aug 12 08:28:10 2026
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Mohamed Mediouni <mohamed@unpredictable.fr>
X-Patchwork-Id: 2284248
Return-Path: <qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org>
X-Original-To: incoming@patchwork.ozlabs.org
Delivered-To: patchwork-incoming@legolas.ozlabs.org
Authentication-Results: legolas.ozlabs.org;
	dkim=pass (2048-bit key;
 unprotected) header.d=unpredictable.fr header.i=@unpredictable.fr
 header.a=rsa-sha256 header.s=sig1 header.b=qk7Pl25O;
	dkim-atps=neutral
Authentication-Results: legolas.ozlabs.org;
 spf=pass (sender SPF authorized) smtp.mailfrom=nongnu.org
 (client-ip=209.51.188.17; helo=lists1p.gnu.org;
 envelope-from=qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org;
 receiver=patchwork.ozlabs.org)
Received: from lists1p.gnu.org (lists1p.gnu.org [209.51.188.17])
	(using TLSv1.2 with cipher ECDHE-ECDSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by legolas.ozlabs.org (Postfix) with ESMTPS id 4hKhVW06jzz1xtl
	for <incoming@patchwork.ozlabs.org>; Wed, 12 Aug 2026 18:29:23 +1000 (AEST)
Received: from localhost ([::1] helo=lists1p.gnu.org)
	by lists1p.gnu.org with esmtp (Exim 4.90_1)
	(envelope-from <qemu-devel-bounces@nongnu.org>)
	id 1wu4K5-0007WJ-Ag; Wed, 12 Aug 2026 04:28:29 -0400
Received: from eggs.gnu.org ([2001:470:142:3::10])
 by lists1p.gnu.org with esmtps (TLS1.2:ECDHE_RSA_AES_256_GCM_SHA384:256)
 (Exim 4.90_1) (envelope-from <mohamed@unpredictable.fr>)
 id 1wu4K4-0007Vz-9E
 for qemu-devel@nongnu.org; Wed, 12 Aug 2026 04:28:28 -0400
Received: from pv-2003g-snip6-11.eps.apple.com ([2a01:b747:3003:205::47]
 helo=outbound.pv.icloud.com)
 by eggs.gnu.org with esmtps (TLS1.2:ECDHE_RSA_AES_256_GCM_SHA384:256)
 (Exim 4.90_1) (envelope-from <mohamed@unpredictable.fr>)
 id 1wu4K2-00052D-Qv
 for qemu-devel@nongnu.org; Wed, 12 Aug 2026 04:28:28 -0400
Received: from outbound.pv.icloud.com (unknown [127.0.0.2])
 by p00-icloudmta-asmtp-us-west-1a-100-percent-10 (Postfix) with ESMTPS id
 DB7AE1800966; Wed, 12 Aug 2026 08:28:21 +0000 (UTC)
X-ICL-RepId: 019ff516-37f6-7c02-afc0-8066abb8607c
X-ICL-Out-Info: 
 HUtFAUMHWwJACUgBTUQeDx5WFlZNRAJCTQ1OHVwPUhxEDlYCVRcOVk1UF1sMWg5WMEwWQx9SD1sTTRlRAVJFVAIZHldQQgVCAlwcWg5DClAcXFZVHxcZUU0PWFsIWwQPH0wMUQJCBVZeSQ4dBFQHXQVdVlACWktCBEtFaFwFXBxAF0gdX2pLVhQEDV0dRQ5WGVoMUg9bDhwWS1heBFNWDkIJSgVdD1oAQAFKC1wDRQBHC0kEQARGXx9RGV4IUytHHkkKVgleCEYRWxRWQ1EZDFBNAUMICghHA00XXjJTBF8RUBY=
Dkim-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=unpredictable.fr;
 s=sig1; t=1786523302; x=1789115302;
 bh=6NRueJblP8aIw/OsQ8F5jg3sZUd49Zglh8TaSrbigVU=;
 h=From:To:Subject:Date:Message-ID:MIME-Version:x-icloud-hme;
 b=qk7Pl25OKC2Jm9Ua4Ow1QxbJAz1NeuucVoSejJCcp3/rdzHeb5NOM2ccJwtJYo24hVjSyU1gZiM+NBDKsIABcvbBZ89T95pyaCgua1SKxMbK3JPrki7UXp32mHnUzTKB2ZlFttJilkcpX7dy2U5QFoc+t1LvFsEzwMXMp40liNd2wlSL9Xxx9EakV8RYT//3yvnrgiJwyyh8sGTrHz43fPBQRXVLxNeva7wAUKioguvqyRTgFnlmVVToSNnn5suCGaOVzPsIXURMWd9o7MyUGReezZL7epn1rmWhOuNeIbTl34FkZOi8RzjdkXyLKzhxCTqhmZysBbek6IVd6QOGQQ==
mail-alias-created-date: 1752046281608
Received: from localhost.localdomain (unknown [17.56.9.36])
 by p00-icloudmta-asmtp-us-west-1a-100-percent-10 (Postfix) with ESMTPSA id
 7C5F61800949; Wed, 12 Aug 2026 08:28:20 +0000 (UTC)
From: Mohamed Mediouni <mohamed@unpredictable.fr>
To: qemu-devel@nongnu.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 "Michael S. Tsirkin" <mst@redhat.com>,
 Pedro Barbuda <pbarbuda@microsoft.com>,
 Mohamed Mediouni <mohamed@unpredictable.fr>,
 Richard Henderson <richard.henderson@linaro.org>
Subject: [PATCH 2/6] hw/i386: fw_cfg: do not set VMX feature control on WHPX
Date: Wed, 12 Aug 2026 10:28:10 +0200
Message-ID: <20260812082814.27217-3-mohamed@unpredictable.fr>
X-Mailer: git-send-email 2.54.0
In-Reply-To: <20260812082814.27217-1-mohamed@unpredictable.fr>
References: <20260812082814.27217-1-mohamed@unpredictable.fr>
MIME-Version: 1.0
X-Proofpoint-GUID: Pix-INabtEUbn_CgeZDyINTiM_8mP7iY
X-Proofpoint-ORIG-GUID: Pix-INabtEUbn_CgeZDyINTiM_8mP7iY
X-Proofpoint-Spam-Details-Enc: AW1haW4tMjYwODEyMDA2NyBTYWx0ZWRfXyPEwbtJ/8aUn
 EqUcoGZcJyUEhfbWz2R+7W62L64xdy8vI3q7fcckkA6Z9PlqeIoc9473X5Zd2E5+lQM/UVxwV+B
 U3TlaV33TUBetKZrYxApbaXPSv9suEBNbTkrZikI9MnO2Zb/jABcw/Iw+cYqoCsP2Ptv+Qny/DZ
 rIDfugPiXtpbNwRt6Ympl4rW4NqjByuNTbYQ9hHv2TQgmR452XEJEhahiLY84Tq+Sh1aaf8YmBs
 Y7sgHEhIkCYEkHLYNSsaYfCFktfmA8F9l/2ysinEEAHIW2vbmzBOowrB2ijEyt5I5tH+gTUpl55
 TGe63XqvrkB78MK+f9WYsMWGg6GnwTGK+RTC3meTP62Nenjxo8OXW/GEn5PIwY=
X-Authority-Info-Out: v=2.4 cv=DqBbOW/+ c=1 sm=1 tr=0 ts=6a7c2ea6
 cx=c_apl:c_pps:t_out a=azHRBMxVc17uSn+fyuI/eg==:117
 a=azHRBMxVc17uSn+fyuI/eg==:17 a=Sv0fKeRqtYgA:10 a=VkNPw1HP01LnGYTKEx00:22
 a=03zX1AGEHLO9id_vLjQA:9
X-JNJ: 
 AAAAAAABYoX05bFN98HoRDILnaQuVvxX5V3Vk4Xy7jErPBGp9Gwx6S1SBV7p656b21NBGnFEAmmionMgYxBPFuYNCvNz8rnY0kcK5Vncg+g0jypL2lZd6nHdQdCgFVfReKUo8svaJ4I/Zbg507t7SgEAqk4uE1tQR4N7TWQsny9O42AkXg3ljV0uzOe5Xl3Rodkhd9Omg+VgXd70cysb5cJx6niItH/bvG7D4VbA7wD97zZt2zVrY3lXKRgkibgSVdo4+D7dqcILnJNiWCRx6GOix3USnKqhhRj1O5gOXovhLoW26AjtMsUfbIa1AX7o+66i0Cmljk5Ey8xEKC58jWnp4ghzzQAlOm9kfJPWgXax0VHgaUVyDa7E2R6ZuGzUTpKb76ZH7IHtUVZAomaLD1H3lDOqgNB7REmAUYpVxpz8xYyhrxJMQipNC1yRiVn9VksGzKAxZrj9xT0QyFmL3EkaSOXYlGdWffGAy6hdatoItGEE5ZRZxY+el1QEdTjaFm9tUR9p+rgkbB+9m3wYjf2i9d5b4oL1bLnBCW6yKHcZZM+WTo7Rkijqm5amN4ePIZi88ypUOemD+/wmejwOeQuN+cUJCBfcL0/hWo/7qTqAiOFUlHybeB4jaUFyZQwy7opEcfpTdpiRmthGSKkS0Mp9Hj2Y34FPCs0gAKO80VsclIet7e1OkEzjdliJ2bh+SDH/3o6YmTSE5x9MjFdpIs3bAVWqrb1sJ4cy
Received-SPF: pass client-ip=2a01:b747:3003:205::47;
 envelope-from=mohamed@unpredictable.fr; helo=outbound.pv.icloud.com
X-Spam_score_int: -20
X-Spam_score: -2.1
X-Spam_bar: --
X-Spam_report: (-2.1 / 5.0 requ) BAYES_00=-1.9, DKIM_SIGNED=0.1,
 DKIM_VALID=-0.1, DKIM_VALID_AU=-0.1, DKIM_VALID_EF=-0.1,
 SPF_HELO_PASS=-0.001,
 SPF_PASS=-0.001 autolearn=ham autolearn_force=no
X-Spam_action: no action
X-BeenThere: qemu-devel@nongnu.org
X-Mailman-Version: 2.1.29
Precedence: list
List-Id: qemu development <qemu-devel.nongnu.org>
List-Unsubscribe: <https://lists.nongnu.org/mailman/options/qemu-devel>,
 <mailto:qemu-devel-request@nongnu.org?subject=unsubscribe>
List-Archive: <https://lists.nongnu.org/archive/html/qemu-devel>
List-Post: <mailto:qemu-devel@nongnu.org>
List-Help: <mailto:qemu-devel-request@nongnu.org?subject=help>
List-Subscribe: <https://lists.nongnu.org/mailman/listinfo/qemu-devel>,
 <mailto:qemu-devel-request@nongnu.org?subject=subscribe>
Errors-To: qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org
Sender: qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org

Hyper-V doesn't allow setting this bit on build 26100.

Signed-off-by: Mohamed Mediouni <mohamed@unpredictable.fr>
---
 hw/i386/fw_cfg.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/hw/i386/fw_cfg.c b/hw/i386/fw_cfg.c
index d422302c1c..5903994a7c 100644
--- a/hw/i386/fw_cfg.c
+++ b/hw/i386/fw_cfg.c
@@ -13,6 +13,8 @@
  */
 
 #include "qemu/osdep.h"
+#include "system/mshv.h"
+#include "system/whpx.h"
 #include "system/numa.h"
 #include "hw/acpi/acpi.h"
 #include "hw/acpi/aml-build.h"
@@ -182,8 +184,15 @@ void fw_cfg_build_feature_control(MachineState *ms, FWCfgState *fw_cfg)
     uint64_t *val;
 
     cpu_x86_cpuid(env, 1, 0, &unused, &unused, &ecx, &edx);
-    if (ecx & CPUID_EXT_VMX) {
-        feature_control_bits |= FEATURE_CONTROL_VMXON_ENABLED_OUTSIDE_SMX;
+
+    /*
+     * Hyper-V in 26100 disallows this bit to be set.
+     * Otherwise a #GP gets raised.
+     */
+    if (!(whpx_enabled())) {
+        if (ecx & CPUID_EXT_VMX) {
+            feature_control_bits |= FEATURE_CONTROL_VMXON_ENABLED_OUTSIDE_SMX;
+        }
     }
 
     if ((edx & (CPUID_EXT2_MCE | CPUID_EXT2_MCA)) ==

From patchwork Wed Aug 12 08:28:11 2026
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Mohamed Mediouni <mohamed@unpredictable.fr>
X-Patchwork-Id: 2284247
Return-Path: <qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org>
X-Original-To: incoming@patchwork.ozlabs.org
Delivered-To: patchwork-incoming@legolas.ozlabs.org
Authentication-Results: legolas.ozlabs.org;
	dkim=pass (2048-bit key;
 unprotected) header.d=unpredictable.fr header.i=@unpredictable.fr
 header.a=rsa-sha256 header.s=sig1 header.b=Io9AfNSA;
	dkim-atps=neutral
Authentication-Results: legolas.ozlabs.org;
 spf=pass (sender SPF authorized) smtp.mailfrom=nongnu.org
 (client-ip=209.51.188.17; helo=lists1p.gnu.org;
 envelope-from=qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org;
 receiver=patchwork.ozlabs.org)
Received: from lists1p.gnu.org (lists1p.gnu.org [209.51.188.17])
	(using TLSv1.2 with cipher ECDHE-ECDSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by legolas.ozlabs.org (Postfix) with ESMTPS id 4hKhV15t9sz1xwH
	for <incoming@patchwork.ozlabs.org>; Wed, 12 Aug 2026 18:28:56 +1000 (AEST)
Received: from localhost ([::1] helo=lists1p.gnu.org)
	by lists1p.gnu.org with esmtp (Exim 4.90_1)
	(envelope-from <qemu-devel-bounces@nongnu.org>)
	id 1wu4K4-0007WF-Vn; Wed, 12 Aug 2026 04:28:29 -0400
Received: from eggs.gnu.org ([2001:470:142:3::10])
 by lists1p.gnu.org with esmtps (TLS1.2:ECDHE_RSA_AES_256_GCM_SHA384:256)
 (Exim 4.90_1) (envelope-from <mohamed@unpredictable.fr>)
 id 1wu4K3-0007VR-8y
 for qemu-devel@nongnu.org; Wed, 12 Aug 2026 04:28:27 -0400
Received: from pv-2003d-snip6-11.eps.apple.com ([2a01:b747:3003:205::29]
 helo=outbound.pv.icloud.com)
 by eggs.gnu.org with esmtps (TLS1.2:ECDHE_RSA_AES_256_GCM_SHA384:256)
 (Exim 4.90_1) (envelope-from <mohamed@unpredictable.fr>)
 id 1wu4K1-00052L-Hd
 for qemu-devel@nongnu.org; Wed, 12 Aug 2026 04:28:27 -0400
Received: from outbound.pv.icloud.com (unknown [127.0.0.2])
 by p00-icloudmta-asmtp-us-west-1a-100-percent-10 (Postfix) with ESMTPS id
 903831800947; Wed, 12 Aug 2026 08:28:23 +0000 (UTC)
X-ICL-RepId: 019ff516-3e85-7720-a7ae-e673de3dc962
X-ICL-Out-Info: 
 HUtFAUMHWwJACUgBTUQeDx5WFlZNRAJCTQ1OHVwPUhxEDlYCVRcOVk1UF1sMWg5WMEwWQx9SD1sTTRlRAVJFVAIZHldQQgVCAlwcWg5DClAcXFZVHxcZUU0PWFsIWwQPH0wMUQJCBVZeSQ4dBFQHXQVdVlACWktCBEtFaFwFXBxAF0gdX2pLVhQEDV0dRQ5WGVoMUg9bDhwWS1heBFNWDkIJSgVdD1oAQAFKC1wDRQBHC0kEQANGXx9RGV4IUytHHkkKVgleCEYRWxRWQ1EZDFBNAUMICghHA00XXjJTBF8RUBY=
Dkim-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=unpredictable.fr;
 s=sig1; t=1786523304; x=1789115304;
 bh=Ke2xrkwmgPCgpejUY5i9tffDmcMSLiMUb65iOIu/Chw=;
 h=From:To:Subject:Date:Message-ID:MIME-Version:x-icloud-hme;
 b=Io9AfNSAPUoZt+13hx2QErfdE0THv07FSD2PJWWYSO6uzoZjQaDmwm5H9WvPjZ346GdKbmdEalrsiytQw5xxqzI8mjt7bWR6bhECOIIqFYJ299Eyj8mz6Zukhr4r6ma2XKTwu7Ic/Wv9ixIUqUYDaxCS9R7QIDuCl7JSFQj+IQH46mo5jOW8HxCfl3d2mzwC+XrRagO+5wcaC/kM1XROsf7R9xG/vGWnkQuBYUjNimHM9oUdVDrbk7xfHH8OvolcgPNw4tHJkYRP2BnFPv3mc3BRXR0fTJJIWRD3y46rmnoFlPcIYtGOnEJUttW1mOO+RESHi2yErZ39vyNYLVzBJw==
mail-alias-created-date: 1752046281608
Received: from localhost.localdomain (unknown [17.56.9.36])
 by p00-icloudmta-asmtp-us-west-1a-100-percent-10 (Postfix) with ESMTPSA id
 0C1B21800967; Wed, 12 Aug 2026 08:28:21 +0000 (UTC)
From: Mohamed Mediouni <mohamed@unpredictable.fr>
To: qemu-devel@nongnu.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 "Michael S. Tsirkin" <mst@redhat.com>,
 Pedro Barbuda <pbarbuda@microsoft.com>,
 Mohamed Mediouni <mohamed@unpredictable.fr>,
 Richard Henderson <richard.henderson@linaro.org>
Subject: [PATCH 3/6] whpx: i386: enable fast hypercall output
Date: Wed, 12 Aug 2026 10:28:11 +0200
Message-ID: <20260812082814.27217-4-mohamed@unpredictable.fr>
X-Mailer: git-send-email 2.54.0
In-Reply-To: <20260812082814.27217-1-mohamed@unpredictable.fr>
References: <20260812082814.27217-1-mohamed@unpredictable.fr>
MIME-Version: 1.0
X-Proofpoint-Spam-Details-Enc: AW1haW4tMjYwODEyMDA2NyBTYWx0ZWRfXzmXgBdy6K9z2
 H69I6O6R8Br0b1YuvAROQ1kZIv+7sgOEM8dhdXtwZ0IlBFdFzSbTHRox9Jm30q2dSW2gGvkd001
 dGUcx+qSWFjmzr/iByQEGSx/VsZdAmSXiA52D4yEiMly27m7PhCUALWTMR0uE1bHe/e+tAasCxT
 RDUmDDIpjE0C1Rx4G8X0oHtqUrk9d/b04WnmRZRoSOmG8ZolfOHxRTNJHKtDrnIYEGY8F//KaZ9
 cTrFu+XKYCz9qeyNgNVX+Hq91JNSDu2kl9FJj0hgqslV4V5YSMp8m3Q4y4F34LM4e1ZaWbeDeY7
 IEaXEfDni9NuPDOR5OPTwxG6eMwC+ixuZHfCXaN+eztwwAr8dOeLkMXwgQaFaM=
X-Proofpoint-GUID: gCvJFNmCPvKV5fSYUjnIRE9VPbXPwG_A
X-Authority-Info-Out: v=2.4 cv=Uvdu9uwB c=1 sm=1 tr=0 ts=6a7c2ea7
 cx=c_apl:c_pps:t_out a=azHRBMxVc17uSn+fyuI/eg==:117
 a=azHRBMxVc17uSn+fyuI/eg==:17 a=Sv0fKeRqtYgA:10 a=VkNPw1HP01LnGYTKEx00:22
 a=__wDYlw7T4YerZGgBbMA:9
X-Proofpoint-ORIG-GUID: gCvJFNmCPvKV5fSYUjnIRE9VPbXPwG_A
X-JNJ: 
 AAAAAAABtpNDyKogb/IorJ4LTiVxxzAcpBWym9GmaJO4LkzHbN65+tQoZrza2+92LcBT3DGWUxAaGe80T6pj613ckqlX10y1Y0e/Lvdk39jXAjAoLezDg835f+9GnpFH5wYpMqaQRPkBX3cwPNeZXY4rRX+sQox4ZwgJpobWA1iivLEm+kR8LYfKv+7d+O49W/NGMSkEp4HeSiPLK0OiuGW5Fs3QuhaS8p9N+iWaVzRwVp/rYMOSPt0kKFNei23fKa4z8VDyAyZVWMzqHhcZ4hKLXnU7v4DXljedPsWS7hUt7pRMS6j4YNUZ1+/UtdRErUZ66PXJVRI3VYKBONh4Sr8KiZAjjRfnkHE7z52Q1miKnQ0BKdGRQC794FEFX8v9uo10Gb+UfNilT4jqjjCMI78vPrG4v6umfN9mB6SwAfQfaIP3RVZJUmZYmlt5TI1MxlKzfuQwwsA6aBwDs6A/aLBkCDLNRv5Oo2Cj5IuFFgQCQnBDr9qoyGxDkD/7wMkD3dfygay8Fcl8HVOCzIaa9t3JW6fP7EkFhA2RRi2Kwmq53N1TkrHYxtGrkOvL8iBVd13CWY/JGHrdIdxzMZhY3KtYWxT5NmS9yogZ8UJYpSTybBgXZKOXpOYANZAzh3ImqsZOTUgFcFmr8p9gjlDW2StzCfbQ3kVDkOMoNAsdo3ZnSB3dnd09fymH4YBusjn0LPxJODmCCeYhpiY2QA==
Received-SPF: pass client-ip=2a01:b747:3003:205::29;
 envelope-from=mohamed@unpredictable.fr; helo=outbound.pv.icloud.com
X-Spam_score_int: -20
X-Spam_score: -2.1
X-Spam_bar: --
X-Spam_report: (-2.1 / 5.0 requ) BAYES_00=-1.9, DKIM_SIGNED=0.1,
 DKIM_VALID=-0.1, DKIM_VALID_AU=-0.1, DKIM_VALID_EF=-0.1,
 SPF_HELO_PASS=-0.001,
 SPF_PASS=-0.001 autolearn=ham autolearn_force=no
X-Spam_action: no action
X-BeenThere: qemu-devel@nongnu.org
X-Mailman-Version: 2.1.29
Precedence: list
List-Id: qemu development <qemu-devel.nongnu.org>
List-Unsubscribe: <https://lists.nongnu.org/mailman/options/qemu-devel>,
 <mailto:qemu-devel-request@nongnu.org?subject=unsubscribe>
List-Archive: <https://lists.nongnu.org/archive/html/qemu-devel>
List-Post: <mailto:qemu-devel@nongnu.org>
List-Help: <mailto:qemu-devel-request@nongnu.org?subject=help>
List-Subscribe: <https://lists.nongnu.org/mailman/listinfo/qemu-devel>,
 <mailto:qemu-devel-request@nongnu.org?subject=subscribe>
Errors-To: qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org
Sender: qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org

Already enabled on Arm so extending it to x86.

Signed-off-by: Mohamed Mediouni <mohamed@unpredictable.fr>
---
 target/i386/whpx/whpx-all.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/target/i386/whpx/whpx-all.c b/target/i386/whpx/whpx-all.c
index 634d542821..1800ba8eb1 100644
--- a/target/i386/whpx/whpx-all.c
+++ b/target/i386/whpx/whpx-all.c
@@ -3247,6 +3247,7 @@ int whpx_accel_init(AccelState *as, MachineState *ms)
 
     synthetic_features.Bank0.HypervisorPresent = 1;
     synthetic_features.Bank0.Hv1 = 1;
+    synthetic_features.Bank0.FastHypercallOutput = 1;
     synthetic_features.Bank0.AccessVpRunTimeReg = 1;
     synthetic_features.Bank0.AccessPartitionReferenceCounter = 1;
     synthetic_features.Bank0.AccessPartitionReferenceTsc = 1;

From patchwork Wed Aug 12 08:28:12 2026
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Mohamed Mediouni <mohamed@unpredictable.fr>
X-Patchwork-Id: 2284250
Return-Path: <qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org>
X-Original-To: incoming@patchwork.ozlabs.org
Delivered-To: patchwork-incoming@legolas.ozlabs.org
Authentication-Results: legolas.ozlabs.org;
	dkim=pass (2048-bit key;
 unprotected) header.d=unpredictable.fr header.i=@unpredictable.fr
 header.a=rsa-sha256 header.s=sig1 header.b=Q2HgJeLw;
	dkim-atps=neutral
Authentication-Results: legolas.ozlabs.org;
 spf=pass (sender SPF authorized) smtp.mailfrom=nongnu.org
 (client-ip=209.51.188.17; helo=lists1p.gnu.org;
 envelope-from=qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org;
 receiver=patchwork.ozlabs.org)
Received: from lists1p.gnu.org (lists1p.gnu.org [209.51.188.17])
	(using TLSv1.2 with cipher ECDHE-ECDSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by legolas.ozlabs.org (Postfix) with ESMTPS id 4hKhVp6ZY6z1xtl
	for <incoming@patchwork.ozlabs.org>; Wed, 12 Aug 2026 18:29:38 +1000 (AEST)
Received: from localhost ([::1] helo=lists1p.gnu.org)
	by lists1p.gnu.org with esmtp (Exim 4.90_1)
	(envelope-from <qemu-devel-bounces@nongnu.org>)
	id 1wu4K8-0007Xw-4f; Wed, 12 Aug 2026 04:28:32 -0400
Received: from eggs.gnu.org ([2001:470:142:3::10])
 by lists1p.gnu.org with esmtps (TLS1.2:ECDHE_RSA_AES_256_GCM_SHA384:256)
 (Exim 4.90_1) (envelope-from <mohamed@unpredictable.fr>)
 id 1wu4K6-0007XR-S0
 for qemu-devel@nongnu.org; Wed, 12 Aug 2026 04:28:30 -0400
Received: from pv-2003l-snip6-10.eps.apple.com ([2a01:b747:3003:205::78]
 helo=outbound.pv.icloud.com)
 by eggs.gnu.org with esmtps (TLS1.2:ECDHE_RSA_AES_256_GCM_SHA384:256)
 (Exim 4.90_1) (envelope-from <mohamed@unpredictable.fr>)
 id 1wu4K5-00052y-Av
 for qemu-devel@nongnu.org; Wed, 12 Aug 2026 04:28:30 -0400
Received: from outbound.pv.icloud.com (unknown [127.0.0.2])
 by p00-icloudmta-asmtp-us-west-1a-100-percent-10 (Postfix) with ESMTPS id
 83A4D180096A; Wed, 12 Aug 2026 08:28:25 +0000 (UTC)
X-ICL-RepId: 019ff516-45e3-755e-a964-dd6a1574f0bd
X-ICL-Out-Info: 
 HUtFAUMHWwJACUgBTUQeDx5WFlZNRAJCTQ1OHVwPUhxEDlYCVRcOVk1UF1sMWg5WMEwWQx9SD1sTTRlRAVJFVAIZHldQQgVCAlwcWg5DClAcXFZVHxcZUU0PWFsIWwQPH0wMUQJCBVZeSQ4dBFQHXQVdVlACWktCBEtFaFwFXBxAF0gdX2pLVhQEDV0dRQ5WGVoMUg9bDhwWS1heBFNWDkIJSgVdD1oAQAFKC1wDRQBHC0kEQAJGXx9RGV4IUytHHkkKVgleCEYRWxRWQ1EZDFBNAUMICghHA00XXjJTBF8RUBY=
Dkim-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=unpredictable.fr;
 s=sig1; t=1786523308; x=1789115308;
 bh=UmbgY7zx2YTbIb9NF7G0shV9RsqD0iTOLo1LbE0fZvU=;
 h=From:To:Subject:Date:Message-ID:MIME-Version:x-icloud-hme;
 b=Q2HgJeLwI+OF/gdjCnV4avjPVyXDLxG96vreArkta2bKOqvw4siNEfvKLpgmrzpQ696jgFnoraZXorrBRiibcs4peD67daLCu0R3IInkdMUvrDCBJHeG/KhoTQ2LxsCvZxoGKwyNk0AdyF2/JgXiSPYgx/ZIuSrJaoeEw7yNgiV1u1vnCMLj0/6m4ll4n+ilwr22lFcDxC8xMjFQXwQUG+sBR9kw4DZ5nh6mByUKG9wqBDgpCm39uUOn9LSepR5ctAHdL7Xx6w32DMs2AN1GcLsVrx6UiQgZ32fDhZ+XWU4v6/HKvkTLWBgBErDCcKf4Oj3srsKTKgUlheV6SI/NDg==
mail-alias-created-date: 1752046281608
Received: from localhost.localdomain (unknown [17.56.9.36])
 by p00-icloudmta-asmtp-us-west-1a-100-percent-10 (Postfix) with ESMTPSA id
 989621800950; Wed, 12 Aug 2026 08:28:23 +0000 (UTC)
From: Mohamed Mediouni <mohamed@unpredictable.fr>
To: qemu-devel@nongnu.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 "Michael S. Tsirkin" <mst@redhat.com>,
 Pedro Barbuda <pbarbuda@microsoft.com>,
 Mohamed Mediouni <mohamed@unpredictable.fr>,
 Richard Henderson <richard.henderson@linaro.org>
Subject: [PATCH 4/6] whpx: i386: synchronise PAT too
Date: Wed, 12 Aug 2026 10:28:12 +0200
Message-ID: <20260812082814.27217-5-mohamed@unpredictable.fr>
X-Mailer: git-send-email 2.54.0
In-Reply-To: <20260812082814.27217-1-mohamed@unpredictable.fr>
References: <20260812082814.27217-1-mohamed@unpredictable.fr>
MIME-Version: 1.0
X-Proofpoint-GUID: JTnvClEPrnBC1RQLsnUEqwn0rnt6Rj20
X-Authority-Info-Out: v=2.4 cv=VtYuwu2n c=1 sm=1 tr=0 ts=6a7c2eaa
 cx=c_apl:c_pps:t_out a=azHRBMxVc17uSn+fyuI/eg==:117
 a=azHRBMxVc17uSn+fyuI/eg==:17 a=Sv0fKeRqtYgA:10 a=VkNPw1HP01LnGYTKEx00:22
 a=NEAV23lmAAAA:8 a=pUpLrfO9kievb6knzj0A:9
X-Proofpoint-Spam-Details-Enc: AW1haW4tMjYwODEyMDA2NyBTYWx0ZWRfX5/V52c2xYpJk
 SD9rKIz6K4pDX9qzOdBpeu++rz7gZ0QxLwKhjB2NKDv0s+uloDCZay9Brwi53RHvDbvDxrDOiEL
 qWUJTNEfZKFM/TQDEbllQyt+js974gRbfOE8ZQTpGhTxiIHfrVNuXgCvHp3y/TJspYP7M3Ti2SW
 +jXVuHvyVY91Ohm7DAmdWDpocIGA2ahhWKN7KlB6F/2RedlP7snXWa6n4erPOFavVRP43LF/iGn
 YlJePawCP06Sv0G9r5w8VLH7U/A5hpu0NFu523+3FBDXqtWuT0wS9hijj4+bMpEPeCqEs1mBqiQ
 OH4UOXkR26nkZlunYDURjgqjKcg66a/HmLWP6HqRF0B90tdTPR1lO1dG4zoTgw=
X-Proofpoint-ORIG-GUID: JTnvClEPrnBC1RQLsnUEqwn0rnt6Rj20
X-JNJ: 
 AAAAAAABkhA89YGtKz2+Sp1TAcqiAByNvZHdp5hgPhKOmX8kkMmLzXxzN2oRbwb2Nzn26DizaLj0ANh1pS+ABsOuDmvwvMzEmGmWf6PRmbEs6gF4qxbfzWCRd2NRsDdO89IkNgX9fGysYRA9IawF+S7vPbhjecvccoiCWJnPn8r3d+MQDplRDQ1xBbEgo1xHMQi17dXbUroTRwAvjQ4x2qPSR/cZwjjDiQpROqqbz5pRJVgBm/k11Ietr7BcQCcYv4xizu6PiK+JzPFYTjenvMgISOCIXolBqaaxyu257504vQ/6nJWjqlQNDmSuk0V+dJV2v+U6GOeDuqL7IuSFb1F0nliRKLcWf79DUKgIqHi0ZVapqPuG9jWpZywMsb8qrgxk7nGHnOKcrlV5zd143/fhj2QxXWiabibHqt3PLgsAge1ExXeVscZKA1QWywznTWnVIaFP1TpOVGPmuvYoBH+feUygG28FIkQSNcLeGO1NMOyYaZfkxrnNo/ETfp6GiW+nOYgQTlonRhnB67oP7u9qbUQs9hcrNP+Un+dOw4HhRp+Z2+36wlMHesmkW3icZu7tLePc6qA69+2/6UWUNa4dRU1JVi+FNIkVj0hO95j6rlDPEv2zvAag1SEsJU4DiVkdFXazdueiu9qDkz6sSvp1369VX61Ic8uXio0UUZ9IoY+tOeT+wkJI1WO/ImqB53HknbcpEz6doh2Mpurg7T/xsDIc+45iC508LaRJd0hb
Received-SPF: pass client-ip=2a01:b747:3003:205::78;
 envelope-from=mohamed@unpredictable.fr; helo=outbound.pv.icloud.com
X-Spam_score_int: -20
X-Spam_score: -2.1
X-Spam_bar: --
X-Spam_report: (-2.1 / 5.0 requ) BAYES_00=-1.9, DKIM_SIGNED=0.1,
 DKIM_VALID=-0.1, DKIM_VALID_AU=-0.1, DKIM_VALID_EF=-0.1,
 SPF_HELO_PASS=-0.001,
 SPF_PASS=-0.001 autolearn=ham autolearn_force=no
X-Spam_action: no action
X-BeenThere: qemu-devel@nongnu.org
X-Mailman-Version: 2.1.29
Precedence: list
List-Id: qemu development <qemu-devel.nongnu.org>
List-Unsubscribe: <https://lists.nongnu.org/mailman/options/qemu-devel>,
 <mailto:qemu-devel-request@nongnu.org?subject=unsubscribe>
List-Archive: <https://lists.nongnu.org/archive/html/qemu-devel>
List-Post: <mailto:qemu-devel@nongnu.org>
List-Help: <mailto:qemu-devel-request@nongnu.org?subject=help>
List-Subscribe: <https://lists.nongnu.org/mailman/listinfo/qemu-devel>,
 <mailto:qemu-devel-request@nongnu.org?subject=subscribe>
Errors-To: qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org
Sender: qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org

https://github.com/cmspam/winq-emu hints that
this might be wanted for some use-cases and
it's a step towards full state sync.

Signed-off-by: Mohamed Mediouni <mohamed@unpredictable.fr>
---
 target/i386/whpx/whpx-all.c | 12 +++++-------
 1 file changed, 5 insertions(+), 7 deletions(-)

diff --git a/target/i386/whpx/whpx-all.c b/target/i386/whpx/whpx-all.c
index 1800ba8eb1..2739a80da3 100644
--- a/target/i386/whpx/whpx-all.c
+++ b/target/i386/whpx/whpx-all.c
@@ -115,7 +115,7 @@ static const WHV_REGISTER_NAME whpx_register_names[] = {
 #ifdef TARGET_X86_64
     WHvX64RegisterKernelGsBase,
 #endif
-    /* WHvX64RegisterPat, */
+    WHvX64RegisterPat,
     WHvX64RegisterSysenterCs,
     WHvX64RegisterSysenterEip,
     WHvX64RegisterSysenterEsp,
@@ -624,9 +624,8 @@ void whpx_set_registers(CPUState *cpu, WHPXStateLevel level)
         assert(whpx_register_names[idx] == WHvX64RegisterKernelGsBase);
         vcxt.values[idx++].Reg64 = env->kernelgsbase;
 #endif
-
-        /* WHvX64RegisterPat - Skipped */
-
+        assert(whpx_register_names[idx] == WHvX64RegisterPat);
+        vcxt.values[idx++].Reg64 = env->pat;
         assert(whpx_register_names[idx] == WHvX64RegisterSysenterCs);
         vcxt.values[idx++].Reg64 = env->sysenter_cs;
         assert(whpx_register_names[idx] == WHvX64RegisterSysenterEip);
@@ -962,9 +961,8 @@ void whpx_get_registers(CPUState *cpu, WHPXStateLevel level)
     assert(whpx_register_names[idx] == WHvX64RegisterKernelGsBase);
     env->kernelgsbase = vcxt.values[idx++].Reg64;
 #endif
-
-    /* WHvX64RegisterPat - Skipped */
-
+    assert(whpx_register_names[idx] == WHvX64RegisterPat);
+    env->pat = vcxt.values[idx++].Reg64;
     assert(whpx_register_names[idx] == WHvX64RegisterSysenterCs);
     env->sysenter_cs = vcxt.values[idx++].Reg64;
     assert(whpx_register_names[idx] == WHvX64RegisterSysenterEip);

From patchwork Wed Aug 12 08:28:13 2026
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Mohamed Mediouni <mohamed@unpredictable.fr>
X-Patchwork-Id: 2284249
Return-Path: <qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org>
X-Original-To: incoming@patchwork.ozlabs.org
Delivered-To: patchwork-incoming@legolas.ozlabs.org
Authentication-Results: legolas.ozlabs.org;
	dkim=pass (2048-bit key;
 unprotected) header.d=unpredictable.fr header.i=@unpredictable.fr
 header.a=rsa-sha256 header.s=sig1 header.b=GX4tBWKP;
	dkim-atps=neutral
Authentication-Results: legolas.ozlabs.org;
 spf=pass (sender SPF authorized) smtp.mailfrom=nongnu.org
 (client-ip=209.51.188.17; helo=lists1p.gnu.org;
 envelope-from=qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org;
 receiver=patchwork.ozlabs.org)
Received: from lists1p.gnu.org (lists1p.gnu.org [209.51.188.17])
	(using TLSv1.2 with cipher ECDHE-ECDSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by legolas.ozlabs.org (Postfix) with ESMTPS id 4hKhVk697xz1xtl
	for <incoming@patchwork.ozlabs.org>; Wed, 12 Aug 2026 18:29:34 +1000 (AEST)
Received: from localhost ([::1] helo=lists1p.gnu.org)
	by lists1p.gnu.org with esmtp (Exim 4.90_1)
	(envelope-from <qemu-devel-bounces@nongnu.org>)
	id 1wu4K7-0007Xl-P7; Wed, 12 Aug 2026 04:28:31 -0400
Received: from eggs.gnu.org ([2001:470:142:3::10])
 by lists1p.gnu.org with esmtps (TLS1.2:ECDHE_RSA_AES_256_GCM_SHA384:256)
 (Exim 4.90_1) (envelope-from <mohamed@unpredictable.fr>)
 id 1wu4K6-0007XJ-2t
 for qemu-devel@nongnu.org; Wed, 12 Aug 2026 04:28:30 -0400
Received: from pv-2003e-snip6-4.eps.apple.com ([2a01:b747:3003:205::2d]
 helo=outbound.pv.icloud.com)
 by eggs.gnu.org with esmtps (TLS1.2:ECDHE_RSA_AES_256_GCM_SHA384:256)
 (Exim 4.90_1) (envelope-from <mohamed@unpredictable.fr>)
 id 1wu4K4-00052o-M4
 for qemu-devel@nongnu.org; Wed, 12 Aug 2026 04:28:29 -0400
Received: from outbound.pv.icloud.com (unknown [127.0.0.2])
 by p00-icloudmta-asmtp-us-west-1a-100-percent-10 (Postfix) with ESMTPS id
 9F905180096C; Wed, 12 Aug 2026 08:28:26 +0000 (UTC)
X-ICL-RepId: 019ff516-4a75-7650-9445-fd7ff87a1d22
X-ICL-Out-Info: 
 HUtFAUMHWwJACUgATUQeDx5WFlZNRAJCTQ1OHVwPUhxEDlYCVRcOVk1UF1sMWg5WMEwWQx9SD1sTTRlRAVJFVAIZHldQQgVCAlwcWg5DClAcXFZVHxcZUU0PWFsIWwQPH0wMUQJCBVZeSQ4dBFQHXQVdVlACWktCBEtFaFwFXBxAF0gdX2pLVhQEDV0dRQ5WGVoMUg9bDhwWS1heBFNWDkIJSgVdD1oAQAFKC1wDRQBHC0kEQAFGXx9RGV4IUytHHkkKVgleCEYRWxRWQ1EZDFBNAUMICghHA00XXjJTBF8RUBY=
Dkim-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=unpredictable.fr;
 s=sig1; t=1786523307; x=1789115307;
 bh=IkhvoWWMz0utjxTwPeKdYrtgPwXhsLr2rZ70Y47kU7w=;
 h=From:To:Subject:Date:Message-ID:MIME-Version:x-icloud-hme;
 b=GX4tBWKPZpYZeNLxWt4oraPYEA4bhSrl9u3mdUbr9DhuEN7giXtNI+K+FtIEyjt+pbyNZKLWhEcW9/+1n64rivdPzNHwCGn6sCeX6Lgkr/x90bs3U9xMVKo9s+qeOCNYJYiLJs2x6EwXSOMvANO4sbwCnCqGXY74HdpGqWx4izTmBhuq9qhTtKHInKST6Ic3sAEixlS0b9qrhD23osgMmChXNqmtVmtkpEJpnM+6DupxiovLsDLqMbmUyUh/RCmUJXbG43ah/Bx+/mLT0bEhbgFuvgdZ4/b1M+A2nle5wb07NDCobrpuB/i02wX2JqBDy1J1JqxAVuXDREP3gGOQ4g==
mail-alias-created-date: 1752046281608
Received: from localhost.localdomain (unknown [17.56.9.36])
 by p00-icloudmta-asmtp-us-west-1a-100-percent-10 (Postfix) with ESMTPSA id
 273CC1800969; Wed, 12 Aug 2026 08:28:25 +0000 (UTC)
From: Mohamed Mediouni <mohamed@unpredictable.fr>
To: qemu-devel@nongnu.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 "Michael S. Tsirkin" <mst@redhat.com>,
 Pedro Barbuda <pbarbuda@microsoft.com>,
 Mohamed Mediouni <mohamed@unpredictable.fr>,
 Richard Henderson <richard.henderson@linaro.org>
Subject: [PATCH 5/6] whpx: i386: work around Hyper-V FP state oddities
Date: Wed, 12 Aug 2026 10:28:13 +0200
Message-ID: <20260812082814.27217-6-mohamed@unpredictable.fr>
X-Mailer: git-send-email 2.54.0
In-Reply-To: <20260812082814.27217-1-mohamed@unpredictable.fr>
References: <20260812082814.27217-1-mohamed@unpredictable.fr>
MIME-Version: 1.0
X-Proofpoint-Spam-Details-Enc: AW1haW4tMjYwODEyMDA2NyBTYWx0ZWRfX+A3KnhI6aR9E
 vRvZV+n66oflUNJ/05cwWWWQC6RIzR9PgIG/vQZ3VCPek2P1TZ/rHaVmFq7yI6JHtI38IQk6/2p
 V/kSYCwk/CILmv6j0Wxbm9kY3Ml6nkHt8gu9uP42fKeViblG4djE29vQ4/UW2fpku8xvZSQeACP
 3Q1mXnxPOQyuBk2z7jBAcxMTYxLWU48KHIBoH+OD371dJhyjOlvM1cZPiOTXg6UIa3VZ2pMM5Xv
 ax/VW0HIuFeax68fE4EaaCafYGUfCamIrvI/ysgfoqejjR6S/cMwZrrO1NnBfo900CDq9zjX4X1
 QDkwR2EuzVyuDY/pRAV/OfiQdYE6akNRPi5vKY6fs8kmVRU7BowkcCnobLpOwA=
X-Authority-Info-Out: v=2.4 cv=As3jHe9P c=1 sm=1 tr=0 ts=6a7c2eaa
 cx=c_apl:c_pps:t_out a=azHRBMxVc17uSn+fyuI/eg==:117
 a=azHRBMxVc17uSn+fyuI/eg==:17 a=Sv0fKeRqtYgA:10 a=VkNPw1HP01LnGYTKEx00:22
 a=VRRVvkFWJP6kdw74u38A:9
X-Proofpoint-GUID: Tf0cm1y7ynXG9fLvM2jT0_W2lOp8PqYn
X-Proofpoint-ORIG-GUID: Tf0cm1y7ynXG9fLvM2jT0_W2lOp8PqYn
X-JNJ: 
 AAAAAAABV85ziyyGK5TeqR+kHHPUMaN/VoORk0Uf2LBsVu7ERLLmVBKCqJuwe3AP+nFzadafVTs4CoDYFtc5lOH3LG8vqu7jUcn5Sb5i5wUGiRICpdJJQ1oWKcfB3UkMomnUMmLD3giBKcesc4cavgzKn+aKjDUHpMPD0Nomu0xbubSeX8lROgQ36K1DYR7rpM9qaoF+HLot33WnNCsOWcBPV1LUikKxsmgklxDGY71tpCpVr4gqvQ8suaYW8Upi599m9rflRazvPPSQeUmPSDyCU+phmN7ySG2F/K5pEwO3mf87KDJt+AkAylEQWMElnsgB8YcIhkhjauw6GV/cyoEVLEG9sosgFYld3mgyKXOTzs6HJohDj3CWXtJm7B8DQwfgUws2t73g5iVJxGqSJoq29+YHkxrwA3QFb0HayZWsW9JcIr2sCRmpERJAZTIf7Y99p+1pdLHdA/HaW//mHmH7kiFrEkct8LzpcWxzdvxfwH59E7+w3ZbSKDipSyVrvb81FN4xDyNK7YJ6iuwj+JvMENOQ7Dyb/a4dTyAtCRwzRo5takHUFqBaZDvEqc1q/6SBHOiCm7Gz8opMsrgDqAudYkFFHnlmP0g5qHr218xWcUYTSVUO9HbnVUU2jNTmo3cofLintMfC9jJGX2RRzVPHTza8eq/vOMzNfx3ZJo1Bk0pgX60gFZOzEFPNBn38foFPT3J/kKk21upL6GwRq1uR90kocgLzZ9PtzDwijPD4nLokJ8WeMCf0Qpg34w==
Received-SPF: pass client-ip=2a01:b747:3003:205::2d;
 envelope-from=mohamed@unpredictable.fr; helo=outbound.pv.icloud.com
X-Spam_score_int: -20
X-Spam_score: -2.1
X-Spam_bar: --
X-Spam_report: (-2.1 / 5.0 requ) BAYES_00=-1.9, DKIM_SIGNED=0.1,
 DKIM_VALID=-0.1, DKIM_VALID_AU=-0.1, DKIM_VALID_EF=-0.1,
 SPF_HELO_PASS=-0.001,
 SPF_PASS=-0.001 autolearn=ham autolearn_force=no
X-Spam_action: no action
X-BeenThere: qemu-devel@nongnu.org
X-Mailman-Version: 2.1.29
Precedence: list
List-Id: qemu development <qemu-devel.nongnu.org>
List-Unsubscribe: <https://lists.nongnu.org/mailman/options/qemu-devel>,
 <mailto:qemu-devel-request@nongnu.org?subject=unsubscribe>
List-Archive: <https://lists.nongnu.org/archive/html/qemu-devel>
List-Post: <mailto:qemu-devel@nongnu.org>
List-Help: <mailto:qemu-devel-request@nongnu.org?subject=help>
List-Subscribe: <https://lists.nongnu.org/mailman/listinfo/qemu-devel>,
 <mailto:qemu-devel-request@nongnu.org?subject=subscribe>
Errors-To: qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org
Sender: qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org

On Hyper-V looks like we need to fetch both the legacy
and new state instead of being able to rely on xsave.

Signed-off-by: Mohamed Mediouni <mohamed@unpredictable.fr>
---
 target/i386/whpx/whpx-all.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/target/i386/whpx/whpx-all.c b/target/i386/whpx/whpx-all.c
index 2739a80da3..56a8b16749 100644
--- a/target/i386/whpx/whpx-all.c
+++ b/target/i386/whpx/whpx-all.c
@@ -614,9 +614,9 @@ void whpx_set_registers(CPUState *cpu, WHPXStateLevel level)
 
         if (whpx_is_xsave_enabled(cpu)) {
             whpx_set_xsave_state(cpu);
-        } else {
-            whpx_set_legacy_fp_registers(cpu, level);
         }
+        whpx_set_legacy_fp_registers(cpu, level);
+
         /* MSRs */
         assert(whpx_register_names[idx] == WHvX64RegisterEfer);
         vcxt.values[idx++].Reg64 = env->efer;
@@ -950,9 +950,8 @@ void whpx_get_registers(CPUState *cpu, WHPXStateLevel level)
 
     if (whpx_is_xsave_enabled(cpu)) {
         whpx_get_xsave_state(cpu);
-    } else {
-        whpx_get_legacy_fp_registers(cpu, level);
     }
+    whpx_get_legacy_fp_registers(cpu, level);
 
     /* MSRs */
     assert(whpx_register_names[idx] == WHvX64RegisterEfer);

From patchwork Wed Aug 12 08:28:14 2026
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Mohamed Mediouni <mohamed@unpredictable.fr>
X-Patchwork-Id: 2284246
Return-Path: <qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org>
X-Original-To: incoming@patchwork.ozlabs.org
Delivered-To: patchwork-incoming@legolas.ozlabs.org
Authentication-Results: legolas.ozlabs.org;
	dkim=pass (2048-bit key;
 unprotected) header.d=unpredictable.fr header.i=@unpredictable.fr
 header.a=rsa-sha256 header.s=sig1 header.b=pmWrcjxi;
	dkim-atps=neutral
Authentication-Results: legolas.ozlabs.org;
 spf=pass (sender SPF authorized) smtp.mailfrom=nongnu.org
 (client-ip=209.51.188.17; helo=lists1p.gnu.org;
 envelope-from=qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org;
 receiver=patchwork.ozlabs.org)
Received: from lists1p.gnu.org (lists1p.gnu.org [209.51.188.17])
	(using TLSv1.2 with cipher ECDHE-ECDSA-AES256-GCM-SHA384 (256/256 bits))
	(No client certificate requested)
	by legolas.ozlabs.org (Postfix) with ESMTPS id 4hKhV15qLvz1xtl
	for <incoming@patchwork.ozlabs.org>; Wed, 12 Aug 2026 18:28:56 +1000 (AEST)
Received: from localhost ([::1] helo=lists1p.gnu.org)
	by lists1p.gnu.org with esmtp (Exim 4.90_1)
	(envelope-from <qemu-devel-bounces@nongnu.org>)
	id 1wu4KB-0007YK-LG; Wed, 12 Aug 2026 04:28:35 -0400
Received: from eggs.gnu.org ([2001:470:142:3::10])
 by lists1p.gnu.org with esmtps (TLS1.2:ECDHE_RSA_AES_256_GCM_SHA384:256)
 (Exim 4.90_1) (envelope-from <mohamed@unpredictable.fr>)
 id 1wu4KA-0007YC-Qg
 for qemu-devel@nongnu.org; Wed, 12 Aug 2026 04:28:34 -0400
Received: from pv-2003d-snip6-11.eps.apple.com ([2a01:b747:3003:205::29]
 helo=outbound.pv.icloud.com)
 by eggs.gnu.org with esmtps (TLS1.2:ECDHE_RSA_AES_256_GCM_SHA384:256)
 (Exim 4.90_1) (envelope-from <mohamed@unpredictable.fr>)
 id 1wu4K9-00053K-CQ
 for qemu-devel@nongnu.org; Wed, 12 Aug 2026 04:28:34 -0400
Received: from outbound.pv.icloud.com (unknown [127.0.0.2])
 by p00-icloudmta-asmtp-us-west-1a-100-percent-10 (Postfix) with ESMTPS id
 04EBB1800950; Wed, 12 Aug 2026 08:28:28 +0000 (UTC)
X-ICL-RepId: 019ff516-5346-7986-915c-66053f687e6b
X-ICL-Out-Info: 
 HUtFAUMHWwJACUgBTUQeDx5WFlZNRAJCTQ1OHVwPUhxEDlYCVRcOVk1UF1sMWg5WMEwWQx9SD1sTTRlRAVJFVAIZHldQQgVCAlwcWg5DClAcXFZVHxcZUU0PWFsIWwQPH0wMUQJCBVZeSQ4dBFQHXQVdVlACWktCBEtFaFwFXBxAF0gdX2pLVhQEDV0dRQ5WGVoMUg9bDhwWS1heBFNWDkIJSgVdD1oAQAFKC1wDRQBHC0kEQABGXx9RGV4IUytHHkkKVgleCEYRWxRWQ1EZDFBNAUMICghHA00XXjJTBF8RUBY=
Dkim-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=unpredictable.fr;
 s=sig1; t=1786523312; x=1789115312;
 bh=+Yof9uipr19UVMj3HWwA5Hlq6kbuQWRKasISjme4jhc=;
 h=From:To:Subject:Date:Message-ID:MIME-Version:x-icloud-hme;
 b=pmWrcjxiKWWaK1GWrGjr8Nzn3Rq8Ogk5THjOxusjBCnfZvWsHR9IO9kk6YG1PnU5YeDX+/CBSozUIG6Nl8i65m3fXkFP0Lb9rlJtjqAPBcE9Xl6j60xTVuw2lrnPoQjkUp00sS6eHyNa09TQRAK/6T4DKUt81+ynydgeFSDf2U6hEosBiTp7PkX8uUDU7cV+I2+y7xxfSO8HQifmx69eUFEu7bkYzBXO/YviOhVud8xE6LCrgxQL4ii9KUaaZymmWLq6BGRU2gFVuZZ1rMboyID2hMoermFiKMC4Y/Ka97Gz2q6Sde/Yw2oGDarpOG3p0ZiWchbR2KDXL32NLrNcIQ==
mail-alias-created-date: 1752046281608
Received: from localhost.localdomain (unknown [17.56.9.36])
 by p00-icloudmta-asmtp-us-west-1a-100-percent-10 (Postfix) with ESMTPSA id
 AB499180096D; Wed, 12 Aug 2026 08:28:26 +0000 (UTC)
From: Mohamed Mediouni <mohamed@unpredictable.fr>
To: qemu-devel@nongnu.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 "Michael S. Tsirkin" <mst@redhat.com>,
 Pedro Barbuda <pbarbuda@microsoft.com>,
 Mohamed Mediouni <mohamed@unpredictable.fr>,
 Richard Henderson <richard.henderson@linaro.org>
Subject: [PATCH 6/6] whpx: i386: inject back db
Date: Wed, 12 Aug 2026 10:28:14 +0200
Message-ID: <20260812082814.27217-7-mohamed@unpredictable.fr>
X-Mailer: git-send-email 2.54.0
In-Reply-To: <20260812082814.27217-1-mohamed@unpredictable.fr>
References: <20260812082814.27217-1-mohamed@unpredictable.fr>
MIME-Version: 1.0
X-Authority-Info-Out: v=2.4 cv=a8Q9NESF c=1 sm=1 tr=0 ts=6a7c2eae
 cx=c_apl:c_pps:t_out a=azHRBMxVc17uSn+fyuI/eg==:117
 a=azHRBMxVc17uSn+fyuI/eg==:17 a=Sv0fKeRqtYgA:10 a=VkNPw1HP01LnGYTKEx00:22
 a=LC-SUz40srEpHiD94tQA:9
X-Proofpoint-ORIG-GUID: ByMYVFCK12OT1xNDMt6v-LMvuCe0biSW
X-Proofpoint-GUID: ByMYVFCK12OT1xNDMt6v-LMvuCe0biSW
X-Proofpoint-Spam-Details-Enc: AW1haW4tMjYwODEyMDA2NyBTYWx0ZWRfXy0c+4CU+uOFQ
 ZtIF7rvyo0CGq/VSjijk0n632iHtQmj7hAp0R4Ayu2yddG+UGexhzJsYnGCSQiUb+4yi8UBT6XG
 FT7Dj9mEPsr8cZZ/bCUjq18MAArw1UpgVs5hz6gUMQml9z3RFkWh/90NgimNy4ZSt5q0MA7bvmg
 ImNw8OEAEQsJtLD0U7AlyRDldzVUy39zyeMphLssRkIllKGwUQ7+m6z4l54+Dk2NGtrSuwaJMZ4
 on/tuEV/WthOgl1jdrrtFqnkJHZbnOgYRxdbQtNSbiUmUOLQh/SDCUiLFoKDrjJecjMcb5STKqI
 7uFEpVBXogh3P5g8e1SxuEeLb5FYyBzQu10Cw0Km1pHwv39XKAkvxgGT53cz3I=
X-JNJ: 
 AAAAAAABDtpE4qwwZKCe3aD3T+J/cWjgpqgNL+69RaC16lHEb5IrHveB+Rkp+y0+VojDPYleZ1Gt0n7KltrSxeXS3NBvEDkQR+O6hATSyrK792FgvhF96hwGReuShXDcqW4CucBLXGZquEGfkTFXcNbATxVdY2/W8t+krjVfsL5rq9JqwoFBTeKWU/hvxahvL0sk7LcTWzKoWwYQ9VLzeI/P37N5pfGMa5Z/EltJ/yO7O1cWyJwVL+6Iw1rDf6bU6dRb2l+Qp378y/7AIPq0OD9O4FRisaPAL43JiC3RDlxrgaQ0oVka9Wvr3/BshuEqbPRb76iup3AkJwL0aHvREIWXvu/2uNMNzQAyp31V4s7WR+mJUpDJ21Z0Bc3WioFZPRM0eq9Ew89HWxWwiEo9qrvJCSY1EX8CJ1V30RECFnsHFZBf5RlYklZZL1hXu5iffEE+vkYSqZ+4nnhvAQ5JfVCSICYSVTzt9EwNoDr5AYYtEbL8Q9uY2ED2QGj++rc1rlbyigx1BocENZKPRIkbl+qmZw9O67V21GgAjFHf7flM4FyHZND8iP2F7SVHpoQSU3QpqYVemLDsN4RkyuiqMeHuZC8Nrl45Ny7se+WAU/YPyBSM1nwtL5atws6r/a8qZWPeGvvPj2WgDhi8OqSJOtYlGnaCqtJUrdPflzKptQ6Kemss78HXnMq0PzsGBzDCFYBvx/ClCN2YBSD9Ga1Igyu0
Received-SPF: pass client-ip=2a01:b747:3003:205::29;
 envelope-from=mohamed@unpredictable.fr; helo=outbound.pv.icloud.com
X-Spam_score_int: -20
X-Spam_score: -2.1
X-Spam_bar: --
X-Spam_report: (-2.1 / 5.0 requ) BAYES_00=-1.9, DKIM_SIGNED=0.1,
 DKIM_VALID=-0.1, DKIM_VALID_AU=-0.1, DKIM_VALID_EF=-0.1,
 SPF_HELO_PASS=-0.001,
 SPF_PASS=-0.001 autolearn=ham autolearn_force=no
X-Spam_action: no action
X-BeenThere: qemu-devel@nongnu.org
X-Mailman-Version: 2.1.29
Precedence: list
List-Id: qemu development <qemu-devel.nongnu.org>
List-Unsubscribe: <https://lists.nongnu.org/mailman/options/qemu-devel>,
 <mailto:qemu-devel-request@nongnu.org?subject=unsubscribe>
List-Archive: <https://lists.nongnu.org/archive/html/qemu-devel>
List-Post: <mailto:qemu-devel@nongnu.org>
List-Help: <mailto:qemu-devel-request@nongnu.org?subject=help>
List-Subscribe: <https://lists.nongnu.org/mailman/listinfo/qemu-devel>,
 <mailto:qemu-devel-request@nongnu.org?subject=subscribe>
Errors-To: qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org
Sender: qemu-devel-bounces+incoming=patchwork.ozlabs.org@nongnu.org

When we get a debug exception other than int1,
inject it back to the guest.

Signed-off-by: Mohamed Mediouni <mohamed@unpredictable.fr>
---
 target/i386/whpx/whpx-all.c | 31 ++++++++++++++++++++++++++-----
 1 file changed, 26 insertions(+), 5 deletions(-)

diff --git a/target/i386/whpx/whpx-all.c b/target/i386/whpx/whpx-all.c
index 56a8b16749..113f79afb9 100644
--- a/target/i386/whpx/whpx-all.c
+++ b/target/i386/whpx/whpx-all.c
@@ -1079,6 +1079,31 @@ static void whpx_inject_back_gpf(CPUState *cpu)
     whpx_set_reg(cpu, WHvRegisterPendingEvent, reg);
 }
 
+static void whpx_inject_back_db(CPUState *cpu)
+{
+    WHV_VP_EXCEPTION_CONTEXT *ctx = &cpu->accel->exit_ctx.VpException;
+    WHV_REGISTER_VALUE reg = {};
+
+    if (ctx->ExceptionInfo.SoftwareException) {
+        /* TODO */
+        warn_report("Was asked to inject software exception.");
+        return;
+    }
+
+    if (ctx->ExceptionType != EXCP01_DB) {
+        warn_report("Was asked to inject exception other than debug.");
+        return;
+    }
+
+    reg.ExceptionEvent.EventPending = 1;
+    reg.ExceptionEvent.EventType = WHvX64PendingEventException;
+    reg.ExceptionEvent.DeliverErrorCode = ctx->ExceptionInfo.ErrorCodeValid;
+    reg.ExceptionEvent.Vector = ctx->ExceptionType;
+    reg.ExceptionEvent.ErrorCode = ctx->ErrorCode;
+    reg.ExceptionEvent.ExceptionParameter = ctx->ExceptionParameter;
+    whpx_set_reg(cpu, WHvRegisterPendingEvent, reg);
+}
+
 static void handle_io(CPUState *env, uint16_t port, void *buffer,
                   int direction, int size, int count)
 {
@@ -2646,11 +2671,7 @@ int whpx_vcpu_run(CPUState *cpu)
             } else if ((vcpu->exit_ctx.VpException.ExceptionType ==
                         WHvX64ExceptionTypeDebugTrapOrFault) &&
                        !cpu_single_stepping(cpu)) {
-                /*
-                 * Just finished stepping over a breakpoint, but the
-                 * gdb does not expect us to do single-stepping.
-                 * Don't do anything special.
-                 */
+                whpx_inject_back_db(cpu);
                 cpu->exception_index = EXCP_INTERRUPT;
             } else {
                 /* Another exception or debug event. Report it to GDB. */
